core/src/transaction_scheduler.rs

```diff
@@ -0,0 +1,1631 @@
+    /// 1. Handle incoming packets and prioritization around them.
+    /// 2. Serve scheduler requests and return responses.
+    fn start_event_loop(
+        t_name: &str,
+        scheduler_request_receiver: Receiver<SchedulerRequest>,
+        packet_receiver: Receiver<Vec<PacketBatch>>,
+        scheduled_accounts: Arc<Mutex<AccountLocks>>,
+        _qos_service: QosService,
+        exit: &Arc<AtomicBool>,
+    ) -> JoinHandle<()> {
+        let exit = exit.clone();
+        Builder::new()
+            .name(t_name.to_string())
+            .spawn(move || {
+                let mut unprocessed_packet_batches = UnprocessedPacketBatches::with_capacity(700_000);
+
+                loop {
+                    select! {
+                        recv(packet_receiver) -> maybe_packet_batches => {
+                            match maybe_packet_batches {
+                                Ok(packet_batches) => {
+                                    Self::handle_packet_batches(&mut unprocessed_packet_batches, packet_batches);
+                                }
+                                Err(_) => {
+                                    break;
+                                }
+                            }
+                        }
+                        recv(scheduler_request_receiver) -> maybe_batch_request => {
+                            match maybe_batch_request {
+                                Ok(batch_request) => {
+                                    Self::handle_scheduler_request(&mut unprocessed_packet_batches, &scheduled_accounts, batch_request);
+                                }
+                                Err(_) => {
+                                    break;
+                                }
+                            }
+                        }
+                        default(Duration::from_millis(100)) => {
+                            if exit.load(Ordering::Relaxed) {
+                                break;
+                            }
+                        }
+                    }
+                }
+            })
+            .unwrap()
+    }
+
+    /// Attempts to schedule a transaction to be executed.
+    fn try_schedule(
+        deserialized_packet: &DeserializedPacket,
+        bank: &Arc<Bank>,
+        highest_wl_blocked_account_fees: &mut HashMap<Pubkey, u64>,
+        highest_rl_blocked_account_fees: &mut HashMap<Pubkey, u64>,
+        scheduled_accounts: &Arc<Mutex<AccountLocks>>,
+    ) -> Result<SanitizedTransaction> {
+        let sanitized_tx = Self::transaction_from_deserialized_packet(
+            deserialized_packet.immutable_section(),
+            &bank.feature_set,
+            bank.vote_only_bank(),
+            bank.as_ref(),
+        )
+        .ok_or_else(|| SchedulerError::InvalidSanitizedTransaction)?;
+
+        let priority = deserialized_packet.immutable_section().priority();
+
+        {
+            let mut scheduled_accounts_l = scheduled_accounts.lock().unwrap();
```

Comment:
perhaps can avoid grabbing this lock until absolutely necessary by pulling the `Self::can_lock_accounts(account_locks, writable_keys, readonly_keys)?;` call out of `check_accounts_not_blocked` and have the call-site down below before `Self::lock_accounts(..)`. Maybe squeeze some juice when multiple banking threads request batches

Reply:
agree! this lock will prob be held too long in current code and would need to see how long all the checks take